presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -782,4 +831,61 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
         cacheBuilder = cacheBuilder.maximumSize(maximumSize);
         return cacheBuilder;
     }
+
+    private static class WithIdentity<T>
+    {
+        private final HiveIdentity identity;
+        private final T key;
+
+        public WithIdentity(HiveIdentity identity, T key)
+        {
+            this.identity = requireNonNull(identity, "identity is null");
+            this.key = requireNonNull(key, "key is null");
+        }
+
+        public HiveIdentity getIdentity()
+        {
+            return identity;
+        }
+
+        public T getKey()
+        {
+            return key;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            WithIdentity<?> other = (WithIdentity<?>) o;
+            return Objects.equals(identity, other.identity) &&
+                   Objects.equals(key, other.key);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(identity, key);
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("identity", identity)
+                    .add("key", key)
+                    .toString();
+        }
+    }
+
+    private HiveIdentity updateIdentity(HiveIdentity identity)
+    {
+         // remove identity if not doing impersonation
+        return delegate.isImpersonationEnabled() ? identity : HiveIdentity.none();
```

Comment:
`HiveIdentity.none()` is going to be used as a cache key many times at the same time, so using a singleton would be nice.